File: packages/python/llm_local_client/context/context_node.py
import uuid
from datetime import datetime
from typing import List, Optional, Dict, Any

class Resource:
    def __init__(self, id: str, type: str, content: str, metadata: Optional[Dict[str, Any]] = None):
        self.id = id
        self.type = type
        self.content = content
        self.metadata = metadata or {}

class ContextNode:
    def __init__(self, role: str, content: str, parent_id: Optional[str] = None, resources: Optional[List[Resource]] = None, id: Optional[str] = None):
        self.id = id or str(uuid.uuid4())
        self.role = role
        self.content = content
        self.parent_id = parent_id
        self.resources = resources or []
        self.children_ids: List[str] = []
        self.timestamp = datetime.utcnow()

    def add_child(self, child_id: str):
        if child_id not in self.children_ids:
            self.children_ids.append(child_id)

File: packages/typescript/src/context/ContextNode.ts
export interface Resource {
  id: string;
  type: string; // 'text', 'image', 'pdf', etc.
  content: string; // path or base64 or raw text
  metadata?: Record<string, any>;
}

export class ContextNode {
  public readonly id: string;
  public readonly parentId: string | null;
  public readonly role: 'user' | 'assistant' | 'system';
  public readonly content: string;
  public readonly resources: Resource[];
  public readonly childrenIds: string[];
  public readonly timestamp: Date;

  constructor(
    id: string,
    role: 'user' | 'assistant' | 'system',
    content: string,
    parentId: string | null = null,
    resources: Resource[] = []
  ) {
    this.id = id;
    this.role = role;
    this.content = content;
    this.parentId = parentId;
    this.resources = resources;
    this.childrenIds = [];
    this.timestamp = new Date();
  }

  addChild(childId: string) {
    if (!this.childrenIds.includes(childId)) {
      this.childrenIds.push(childId);
    }
  }
}
